pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+// This file is part of HydraDX.
+
+// Copyright (C) 2020-2022  Intergalactic, Limited (GIB).
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{ensure, pallet_prelude::DispatchResult, traits::Get};
+use hydradx_traits::{OnLiquidityChangeHandler, OnPoolStateChangeHandler};
+use scale_info::TypeInfo;
+use sp_core::MaxEncodedLen;
+use sp_runtime::traits::{AtLeast32BitUnsigned, CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, Zero};
+use sp_runtime::{ArithmeticError, DispatchError, RuntimeDebug};
+
+pub mod weights;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarking;
+
+#[cfg(test)]
+mod tests;
+
+/// Max trade volume limit multiplier of liquidity that can be traded in a block
+pub const MAX_LIMIT_VALUE: u32 = 10_000;
+
+#[derive(Clone, Encode, Decode, RuntimeDebug, MaxEncodedLen, TypeInfo, Eq, PartialEq)]
+#[scale_info(skip_type_params(T))]
+pub struct TradeVolumeLimit<T: Config> {
+	pub amount_in: T::Balance,
```

Comment:
is it necessary to track both amount? wouldn't be easier to have just one signed value?

Reply:
Yes, it's possible to use i128, but trade volume limit will be limited to u64 (in both directions). And the conversion will not be infallible.

Reply:
ok, not worth probably, i would rename fields to be not confused with swap function params